pkg/generators/git.go

```diff
@@ -36,26 +39,81 @@ func (g *GitGenerator) GenerateParams(appSetGenerator *argoprojiov1alpha1.Applic
 		return nil, EmptyAppSetGeneratorError
 	}
 
+	var err error
+	var res []map[string]string
+	if appSetGenerator.Git.Directories != nil {
+		res, err = g.generateParamsForGitDirectories(appSetGenerator)
+	} else if appSetGenerator.Git.Files != nil {
+		res, err = g.generateParamsForGitFiles(appSetGenerator)
+	} else {
+		return nil, EmptyAppSetGeneratorError
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (g *GitGenerator) generateParamsForGitDirectories(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator) ([]map[string]string, error) {
 	allApps, err := g.repos.GetApps(context.TODO(), appSetGenerator.Git.RepoURL, appSetGenerator.Git.Revision)
 	if err != nil {
 		return nil, err
 	}
 
 	log.WithFields(log.Fields{
-		"allAps": allApps,
-		"total": len(allApps),
-		"repoURL": appSetGenerator.Git.RepoURL,
+		"allAps":   allApps,
+		"total":    len(allApps),
+		"repoURL":  appSetGenerator.Git.RepoURL,
 		"revision": appSetGenerator.Git.Revision,
 	}).Info("applications result from the repo service")
 
-	requestedApps := g.filter(appSetGenerator.Git.Directories, allApps)
+	requestedApps := g.filterApps(appSetGenerator.Git.Directories, allApps)
 
-	res := g.generateParams(requestedApps, appSetGenerator)
+	res := g.generateParamsFromApps(requestedApps, appSetGenerator)
 
 	return res, nil
 }
 
-func (g *GitGenerator) filter(Directories []argoprojiov1alpha1.GitDirectoryGeneratorItem, allApps []string) []string {
+func (g *GitGenerator) generateParamsForGitFiles(appSetGenerator *argoprojiov1alpha1.ApplicationSetGenerator) ([]map[string]string, error) {
+	allPaths := []string{}
```

Comment:
This function is going to create a new git client for every file. I am not sure if this is a costly code (does it uses network / IO for it?), but if so, we should probably initialize it only once

Reply:
The `git` checkout call creates a local `tmp` directory, and it gets reused for subsequent calls, until the `tmp` directory is deleted. 

Every `git` `fetch` call will involve network I/O to the git servers, but it will not pull the repo from scratch. So, I think we are good here.